### PR TITLE
Temportal queries

### PR DESCRIPTION
**Pull request by tilley14**
(Edward Wong and Nikolas Tilley)

Added Geo-Spacial and Temporal Query Endpoints to the Open Source DB  
Added History table to the database  
Asset Properties get saved in the History Table

### CODE REVIEW

File: db/api/v1/db/temporal.db.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+ * @param {moment} end_date - A temporal upper bound of the history search
+ * 
+ * @returns {*} rows - the historic properties that meet the serach parameters
+ */
+const temporalSearch = async (geometry, asset_id, sponsor_name, project_name, asset_type_name, start_date, end_date) => {
+    let query = QUERY_HISTORY;
+    let values = [];
+
+    switch (geometry.type) {
+    case 'Circle':
```
Comment (APalumbo97) [inline db/api/v1/db/temporal.db.js]: Cases should indented once more.

File: db/api/v1/db/temporal.db.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    }
+    if ((typeof project_name !== 'undefined') && (project_name !== '')) {
+        values.push(project_name);
+        query += ` AND LOWER(project.name) = LOWER($${values.length})` + ' ';
+    }
+    if ((typeof asset_type_name !== 'undefined') && (asset_type_name !== '')) {
+        values.push(asset_type_name);
+        query += ` AND LOWER(asset_type.name) = LOWER($${values.length})` + ' ';
+    }
+    if((typeof start_date !== 'undefined') || (typeof end_date !== 'undefined')) {
```
Comment (APalumbo97) [inline db/api/v1/db/temporal.db.js]: Missing space after `if`

File: db/api/v1/db/temporal.db.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        const has_start_date = (typeof start_date !== 'undefined');
+        const has_end_date = (typeof end_date !== 'undefined');
+
+        if (has_start_date && has_end_date) {
+            if(start_date.isValid() && end_date.isValid()) {
```
Comment (APalumbo97) [inline db/api/v1/db/temporal.db.js]: Missing space after `if`

File: db/api/v1/db/temporal.db.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                values.push(start_date.format(utils.shared.dateStringFormat()));
+                const idx_start = values.length;
+                values.push(end_date.format(utils.shared.dateStringFormat()));
+                const idx_end = values.length;
+                query += ` AND history.date BETWEEN $${idx_start} AND $${idx_end}` + ' ';
+            }
+
+        } else if (has_start_date) {
+            if(start_date.isValid()) {
```
Comment (APalumbo97) [inline db/api/v1/db/temporal.db.js]: Missing space after `if`

File: db/api/v1/db/temporal.db.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+                values.push(start_date.format(utils.shared.dateStringFormat()));
+                query += ` AND history.date >= $${values.length}` + ' ';
+            }
+        } else if (has_end_date) {
+            if(end_date.isValid()){
```
Comment (APalumbo97) [inline db/api/v1/db/temporal.db.js]: Missing space after `if`

File: db/api/v1/db/temporal.db.js [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+ * 
+ * @param {geometry} geometry - a GeoJson geometry
+ * @param {number} asset_id - A specific asset's id
+ * @param {string} sponsor_name - A sponsor's id
+ * @param {string} project_name - A projects's id
+ * @param {string} asset_type_name - An asset type's id
+ * @param {moment} start_date - A temporal lower bound of the history search
+ * @param {moment} end_date - A temporal upper bound of the history search
+ * 
+ * @returns {*} rows - the historic properties that meet the serach parameters
```
Comment (exw4141) [inline db/api/v1/db/temporal.db.js]: search

File: db/api/v1/routes/validate.js [hunk 7]
```diff
@@ -175,6 +176,110 @@ const parseCoordinates = (coordinateList) => {
...
+
+    if (typeof geometry.type === 'undefined' || !geometry.type)
+        return ParseResult.failure('geometry \'type\' was undefined');
+
+    if (!Array.isArray(geometry.coordinates))
+        return ParseResult.failure('geometry must have \'coordinates\' with at least one [lon, lat] or [[lon, lat], ...]');
+
+    // Types are case sensitive
+    switch (geometry.type) {
+    case 'Circle':
```
Comment (APalumbo97) [inline db/api/v1/routes/validate.js]: Case statements should be indented once more

File: db/api/v1/routes/validate.js [hunk 8]
```diff
@@ -175,6 +176,110 @@ const parseCoordinates = (coordinateList) => {
...
+    case 'Polygon': 
+        return parsePolygonGeometry(geometry);
+    default:
+        return ParseResult.failure('geometry \'type\' is case sensitive and can be (\'Circle\', \'Polygon\')');
+    }
+};
+
+// We are assuming that this is a geometry of type 'Circle' and has coordinates
+const parseCircleGeometry = (geometry) => {
+    if (geometry.coordinates.length != 2)
```
Comment (APalumbo97) [inline db/api/v1/routes/validate.js]: SINGLE LINE `IF` STATMENTS WITHOUT `{}`. SHAME!

File: db/api/v1/routes/validate.js [hunk 9]
```diff
@@ -175,6 +176,110 @@ const parseCoordinates = (coordinateList) => {
...
+        return ParseResult.failure('Circle geometry \'coordinates\' is formatted [lon, lat]');
+    
+    if (typeof geometry.radius === 'undefined' || !geometry.radius)
+        return ParseResult.failure('Circle geometry must have a \'radius\'');
```
Comment (APalumbo97) [inline db/api/v1/routes/validate.js]: SEE LINE `204`

File: db/api/v1/routes/validate.js [hunk 10]
```diff
@@ -175,6 +176,110 @@ const parseCoordinates = (coordinateList) => {
...
+
+
+    const lon = parseFloat(geometry.coordinates[0]);
+    const lat = parseFloat(geometry.coordinates[1]);
+
+    if (isNaN(lon) || isNaN(lat))
```
Comment (APalumbo97) [inline db/api/v1/routes/validate.js]: SEE LINE `204`

File: db/api/v1/routes/validate.js [hunk 11]
```diff
@@ -175,6 +176,110 @@ const parseCoordinates = (coordinateList) => {
...
+        return ParseResult.failure('error parsing coordinate values');
+
+    const radius = parseFloat(geometry.radius);
+
+    if (isNaN(radius))
+        return ParseResult.failure('error parsing radius');
```
Comment (APalumbo97) [inline db/api/v1/routes/validate.js]: SEE LINE `204`

File: db/api/v1/routes/validate.js [hunk 12]
```diff
@@ -175,6 +176,110 @@ const parseCoordinates = (coordinateList) => {
...
+
+    geometry.radius = radius;
+    geometry.coordinates = [lon, lat];
+
+    return ParseResult.success(geometry);
+};
+
+// We are assuming that this is a geometry of type 'Polygon' and has coordinates
+const parsePolygonGeometry = (geometry) => {
+    if (geometry.coordinates.length < 4)
```
Comment (APalumbo97) [inline db/api/v1/routes/validate.js]: SEE LINE `204`

File: db/api/v1/routes/validate.js [hunk 13]
```diff
@@ -175,6 +176,110 @@ const parseCoordinates = (coordinateList) => {
...
+    let lat = 0;
+    let lon = 0;
+
+    var i;
+    var coordArray;
+    var parsedCoordinates = [];
+    for (i = 0; i < geometry.coordinates.length; i++) {
+        coordArray = geometry.coordinates[i];
+
+        if (coordArray.length != 2)
```
Comment (APalumbo97) [inline db/api/v1/routes/validate.js]: SEE LINE `204`

File: db/api/v1/routes/validate.js [hunk 14]
```diff
@@ -175,6 +176,110 @@ const parseCoordinates = (coordinateList) => {
...
+        if (!isNaN(lon) && validLongitude(lon)
+            && !isNaN(lat) && validLatitude(lat)) {
+            parsedCoordinates.push([lon, lat]);
+        }
+        else {
+            return ParseResult.failure('Unable to parse coordinate. Please format coordinate arrays like [lon, lat]');
+        }
+    }
+
+    if (parsedCoordinates[0].toString().localeCompare(parsedCoordinates[parsedCoordinates.length - 1].toString()) !== 0)
```
Comment (APalumbo97) [inline db/api/v1/routes/validate.js]: SEE LINE `204`

File: db/api/v1/routes/validate.js [hunk 15]
```diff
@@ -247,6 +352,30 @@ const parseProjectName = (name) => {
...
 
+/**
+ * Validates a Sponsors's name. Does not allow empty strings.
+ * 
+ * @param {*} name - the name of the sponsor to parse
+ * @returns {ParseResult} parse success with a valid name or a parse failure
+ */
+const parseSponsorName = (name) => {
+    return (parseString(name, MIN_PROJECT_NAME_LENGTH) 
+        ? ParseResult.success(name)
```
Comment (APalumbo97) [inline db/api/v1/routes/validate.js]: Bonus points for the turing operator

File: db/api/v1/utils/db.utils.js [hunk 16]
```diff
@@ -49,13 +49,32 @@ const makeLineStringHelper = (lon, lat) => {
...
  */
 const makeLineString = (coordinatesList) => {
-
     // An array of LINESTRING coordinate pairs 'lon lat'
-    let coordPairs = [];
+    var coordPairs = [];
 
-    coordinatesList.forEach(point => {
+    var point;
+    for(var i = 0; i < coordinatesList.length; i++) {
```
Comment (APalumbo97) [inline db/api/v1/utils/db.utils.js]: Space after `for`

File: db/api/v1/utils/db.utils.js [hunk 17]
```diff
@@ -49,13 +49,32 @@ const makeLineStringHelper = (lon, lat) => {
...
+ * Creates a postgis LINESTRING from an array of GeoJson coordinates
+ * 
+ * @param {Array} coordinates an array of GeoJson coordinates [lon, lat] points
+ * @returns {string} a postgis LINESTRING
+ */
+const makeLineStringFromGeoJsonCoordinates = (coordinates) => {
+    var coordPairs = [];
+
+    var point;
+    for(var i = 0; i < coordinates.length; i++) {
```
Comment (APalumbo97) [inline db/api/v1/utils/db.utils.js]: Space after `for`

File: db/api/v1/routes/validate.js (hunk 15 above)
Comment (exw4141) [reply db/api/v1/routes/validate.js]: ternary

File: db/api/v1/db/__tests__/temporal.db.test.js [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+                    'project_name': 'ProjectName',
+                    'date': '2020-05-20',
+                }
+            }]
+        };
+
+        const results = await temporalSearch(geometry, undefined, undefined, undefined, undefined, undefined, undefined);
+        await expect(results).toEqual(expect.objectContaining(expectedGeoJson));
+    });
+
```
Comment (APalumbo97) [inline db/api/v1/db/__tests__/temporal.db.test.js]: Remove these newlines.

File: db/api/v1/db/temporal.db.js [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+const D_WITHIN = `
+    ST_DWithin(ST_SetSRID(asset.location, 4326)::geography, 
+        ST_SetSRID(ST_MakePoint($1, $2), 4326)::geography,
+        $3)
+`;
+
+const POLYGON_WITHIN = `
+    ST_Within(asset.location, ST_MakePolygon(ST_GeomFromText($1)))
+`;
+/**
```
Comment (APalumbo97) [inline db/api/v1/db/temporal.db.js]: Newline before comment.

File: db/schema/README.md [hunk 20]
```diff
@@ -39,4 +39,132 @@ SET row_security = off;
...
+SET client_encoding = 'UTF8';
+SET standard_conforming_strings = on;
+SELECT pg_catalog.set_config('search_path', '', false);
+SET check_function_bodies = false;
+SET client_min_messages = warning;
+SET row_security = off;
+```
+
+
+## Backuping up the data_type table
```
Comment (exw4141) [inline db/schema/README.md]: Backupping or backing up?

**Pull request merged**